OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+  ///              column wise. See OpenSim::InputItDir for possible values.
+  /// \param allow_missing Allow for missing values. When set to false, this
+  ///                      function will throw if the input iterator fills up 
+  ///                      the last row/col partially. When set to true, 
+  ///                      missing elements are set to SimTK::NaN.
+  ///
+  /// \throws ZeroElements When input-iterator does not produce any elements.
+  ///                      That is first == last.
+  template<typename InputIt>
+  DataTable_(InputIt first,
+             typename std::enable_if<!std::is_integral<InputIt>::value,
+                                     InputIt>::type last,
+             const size_t ndir,
+             const InputItDir dir = ROWWISE,
+             const bool allow_missing = false) :
+    data{static_cast<int>(dir == ROWWISE ? 1    : ndir), 
+         static_cast<int>(dir == ROWWISE ? ndir :    1)},
+    metadata{},
+    col_ind{} {
+    if(!(first != last))
+      throw ZeroElements{"Input iterators produce zero elements."};
+    if(ndir == 0)
+      throw InvalidEntry{"Input argument 'ndir' cannot be zero."};
+
+    int row{0};
+    int col{0};
+    while(first != last) {
+      data.set(row, col, *first);
+      ++first;
+      if(dir == ROWWISE) {
+        ++col;
+        if(col == static_cast<int>(ndir)  && first != last) {
+          col = 0;
+          ++row;
+          data.resizeKeep(data.nrow() + 1, data.ncol());
+        }
+      } else {
+        ++row;
+        if(row == static_cast<int>(ndir) && first != last) {
+          row = 0;
+          ++col;
+          data.resizeKeep(data.nrow(), data.ncol() + 1);
+        }
+      }
+    }
+    if(!allow_missing) {
+      if((dir == ROWWISE && col != data.ncol()) ||
+         (dir == COLWISE && row != data.nrow()))
+        throw InvalidEntry{"Input iterator did not produce enough elements to "
+                           "fill the last row."};
+    }
+  }
+
+  /// Copy constructor. Perform a deep copy of the DataTable_.
+  DataTable_(const DataTable_& dt) :
+    data{dt.data}, metadata{dt.metadata}, col_ind{dt.col_ind} {
+      // Deep copy of metadata.
+      for(auto& elem : metadata)
+        elem.second = elem.second->clone();
+    }
+
+  /// Virtual copy constructor. Perform a deep copy of the DataTable_.
+  std::unique_ptr<AbstractDataTable> clone() const override {
+    return std::unique_ptr<AbstractDataTable>(new DataTable_{*this});
+  }
+
+  /// Copy assignment operator. Perform a deep copy of the DataTable_.
+  DataTable_& operator=(const DataTable_& dt) {
+    // Release metadata memory.
+    for(auto& elem : metadata)
+      delete elem.second;
+
+    data     = dt.data;
+    metadata = dt.metadata;
+    col_ind  = dt.col_ind;
+
+    // Deep copy of metadata.
+    for(auto& elem : metadata)
+      elem.second = elem.second->clone();
+
+    return *this;
+  }
+
+  /// Move constructor.
+  DataTable_(DataTable_&& dt) = default;
```

Comment:
FYI, I believe this will fail to compile in VS2013. Hopefully we can just declare that OpenSim 4.0 requires VS2015. I think that is the plan.